tests/unit/spec/base/stub-attribution.js

```diff
@@ -214,6 +214,38 @@ describe('stub-attribution.js', function() {
 
             expect(Mozilla.StubAttribution.hasValidData(data)).toBeFalsy();
         });
+
+        it('should validate double-encoded attribution data correctly', function() {
+            /* eslint-disable camelcase */
+            const data = {
+                utm_source: 'addons.mozilla.org',
+                utm_medium: 'referral',
+                utm_campaign: 'amo-fx-cta-607454',
+                utm_content: 'rta%25253AdUJsb2NrMEByYXltb25kaGlsbC5uZXQ',
+                referrer: '',
```

Comment:
Shouldn't this parameter be set to `'https://addons.mozilla.org/'` and the matcher used below be `toBeTruthy()` given that the `utm_content` value looks valid?

Reply:
Ok updated

Reply:
Ha, I see now. It shouldn't be possible because AMO wouldn't do that but if that happens for some reasons, the `referer` value should still be checked to avoid the by-pass.

Reply:
Sorry, I just saw the update. I think we need both test cases.

Reply:
I think both makes sense 👍

Reply:
To be clear, the following test case (copied from a previous iteration) is useful for the bypass (sorry, I got confused when I read it the first time):

```js
        it('should validate double-encoded attribution data correctly', function() {
            /* eslint-disable camelcase */
            const data = {
                utm_source: 'addons.mozilla.org',
                utm_medium: 'referral',
                utm_campaign: 'amo-fx-cta-607454',
                utm_content: 'rta%25253AdUJsb2NrMEByYXltb25kaGlsbC5uZXQ',
                referrer: '',
                ua: 'chrome',
                visit_id: GA_VISIT_ID
            };
            /* eslint-enable camelcase */

            expect(Mozilla.StubAttribution.hasValidData(data)).toBeFalsy();
        });
```

In addition, I think we'd need a test case like this one. We should early return when `utm_content` is longer than `120` characters I believe, instead of trying to decode up to 120 chars.

```diff
diff --git a/tests/unit/spec/base/stub-attribution.js b/tests/unit/spec/base/stub-attribution.js
index 156f5a13d..029eb1aa7 100644
--- a/tests/unit/spec/base/stub-attribution.js
+++ b/tests/unit/spec/base/stub-attribution.js
@@ -246,6 +246,22 @@ describe('stub-attribution.js', function() {
 
             expect(Mozilla.StubAttribution.hasValidData(data)).toBeTruthy();
         });
+
+        it('should limit the length of the UTM content', function() {
+            /* eslint-disable camelcase */
+            const data = {
+                utm_source: 'addons.mozilla.org',
+                utm_medium: 'referral',
+                utm_campaign: 'amo-fx-cta-607454',
+                utm_content: 'rta%' + '25'.repeat(58) + '3AdUJsb2NrMEByYXltb25kaGlsbC5uZXQ',
+                referrer: '',
+                ua: 'chrome',
+                visit_id: GA_VISIT_ID
+            };
+            /* eslint-enable camelcase */
+
+            expect(Mozilla.StubAttribution.hasValidData(data)).toBeFalsy();
+        });
     });
 
     describe('isFirefoxNewScene2', function() {
```

Reply:
Ok, maybe we can return early if `utm_content` starts with `rta` and length > 120 chars. We'd still be iterating over any length of string less than that value though, which is the same as we're doing now.

Thoughts?

Reply:
> @willdurand On the question of returning early if `utm_content` > 120 chars. This would imply we drop all stub attribution data (not just RTAMO attribution) if `utm_content` > 120 chars. I'm not quite clear on the impact of doing that, which makes me hesitant to accept the suggestion.

I see. I think the only option here would be to use a more conservative value then (`200` maybe?).

> Ok, maybe we can return early if `utm_content` starts with `rta` and length > 120 chars. We'd still be iterating over any length of string less than that value though, which is the same as we're doing now.

Could we start by checking the `referer` first and, if it is AMO, then we restrict the length of `utm_content`?

Reply:
> I see. I think the only option here would be to use a more conservative value then (200 maybe?).

I'm going to see if I can find out the maximum length of value we support for attribution, perhaps that might make reasoning about this easier.

> Could we start by checking the referer first and, if it is AMO, then we restrict the length of utm_content?

It seems like the usecase we're more concerned about is if the referrer is not AMO, which represents the large majority of attribution data.